src/psbt.c

```diff
@@ -0,0 +1,2120 @@
+        p += varbuff_to_bytes(&type, 1, p);
+        p += varbuff_to_bytes(output->redeem_script, output->redeem_script_len, p);
+    }
+    // Witness script
+    if (output->witness_script) {
+        type = WALLY_PSBT_OUT_WITNESS_SCRIPT;
+        p += varbuff_to_bytes(&type, 1, p);
+        p += varbuff_to_bytes(output->witness_script, output->witness_script_len, p);
+    }
+    // Keypaths
+    if (output->keypaths) {
+        struct keypath_map *keypaths = output->keypaths;
+        for (i = 0; i < keypaths->num_items; ++i) {
+            size_t origin_len, j;
+            struct keypath_item *item = &keypaths->items[i];
+            if (pubkey_is_compressed(item->pubkey)) {
+                p += varint_to_bytes(34, p);
+                *p = WALLY_PSBT_OUT_BIP32_DERIVATION;
+                p++;
+                memcpy(p, item->pubkey, 33);
+                p += 33;
+            } else {
+                p += varint_to_bytes(66, p);
+                *p = WALLY_PSBT_OUT_BIP32_DERIVATION;
+                p++;
+                memcpy(p, item->pubkey, 65);
+                p += 65;
+            }
+
+            origin_len = 4; // Start with 4 bytes for fingerprint
+            origin_len += item->origin.path_len * sizeof(uint32_t);
+            p += varint_to_bytes(origin_len, p);
+
+            memcpy(p, item->origin.fingerprint, 4);
+            p += 4;
+            for (j = 0; j < item->origin.path_len; ++j) {
+                memcpy(p, &item->origin.path[j], sizeof(uint32_t));
+                p += 4;
+            }
+        }
+    }
+    // Unknowns
+    if (output->unknowns) {
+        for (i = 0; i < output->unknowns->num_items; ++i) {
+            struct unknowns_item *unknown = &output->unknowns->items[i];
+            p += varint_to_bytes(unknown->key_len, p);
+            memcpy(p, unknown->key, unknown->key_len);
+            p += unknown->key_len;
+            p += varint_to_bytes(unknown->value_len, p);
+            memcpy(p, unknown->value, unknown->value_len);
+            p += unknown->value_len;
+        }
+    }
+
+    // Separator
+    *p = WALLY_PSBT_SEPARATOR;
+    p++;
+
+    *bytes_written = p - bytes_out;
+
+    return WALLY_OK;
+}
+
+int wally_psbt_to_bytes(
+    const struct wally_psbt *psbt,
+    unsigned char *bytes_out, size_t len,
+    size_t *bytes_written)
+{
+    unsigned char type, *p = bytes_out, *end = bytes_out + len;
+    size_t calc_len, tx_len, i;
+    int ret;
+
+    if (bytes_written) {
+        *bytes_written = 0;
+    }
+
+    ret = psbt_get_length(psbt, &calc_len);
+    if (ret != WALLY_OK) {
+        return ret;
+    }
+    if (calc_len > len) {
+        return WALLY_EINVAL; // Buffer is not big enough
+    }
+
+    // Magic
+    memcpy(p, WALLY_PSBT_MAGIC, 5);
+    p += 5;
+
+    // Global tx
+    type = WALLY_PSBT_GLOBAL_UNSIGNED_TX;
+    p += varbuff_to_bytes(&type, 1, p);
+    ret = wally_tx_get_length(psbt->tx, 0, &tx_len);
+    if (ret != WALLY_OK) {
+  
```

Comment:
buff is always non null if it reaches here

Reply:
And?

Reply:
if the analysis is correct no need for the if statement